### PR TITLE
test: test write methods against more varied layouts

### PR DESCRIPTION
**Pull request by wjones127**
Adds a test utility to generate a more hostile layout. Adds to one test function. Will add to more in follow up PRs.

### CODE REVIEW

File: rust/lance-datafusion/src/utils.rs [hunk 1]
```diff
@@ -46,12 +46,21 @@ pub async fn reader_to_stream(
...
+    Ok((
+        Box::new(reader) as Box<dyn RecordBatchReader + Send>,
+        schema,
+    ))
+}
 
+/// Convert reader to a stream.
+///
+/// The reader will be called in a background thread.
+pub fn reader_to_stream(batches: Box<dyn RecordBatchReader + Send>) -> SendableRecordBatchStream {
```
Comment (wjones127) [inline rust/lance-datafusion/src/utils.rs:59]: Found we sometimes didn't need to peek the reader, so separated this into two different functions: `peek_reader_schema()` and `reader_to_stream()`.

File: rust/lance/src/dataset.rs [hunk 2]
```diff
@@ -3462,32 +3465,33 @@ mod tests {
...
-            ..Default::default()
-        };
-        let mut dataset = Dataset::write(batches, test_uri, Some(write_params))
-            .await
-            .unwrap();
+        // Split over two files.
+        let batches = vec![data.slice(0, 50), data.slice(50, 50)];
+        let mut dataset = TestDatasetGenerator::new(batches)
+            .make_hostile(test_uri)
+            .await;
```
Comment (wjones127) [inline rust/lance/src/dataset.rs:3495]: This is the one test I've integrated this with. In follow up PRs, we can use this in additional tests.

File: rust/lance/src/dataset/fragment/write.rs
Comment (wjones127) [inline rust/lance/src/dataset/fragment/write.rs:1]: The current `Fragment::create()` API didn't allow passing an explicit `lance::Schema`, but we need that to generate layouts with specific field ids. So I moved this all over to a nice builder.

File: rust/lance/src/dataset/fragment/write.rs [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+        let filename = format!("{}.lance", Uuid::new_v4());
+        let mut fragment = Fragment::with_file_legacy(id, &filename, &schema, None);
+        let full_path = base_path.child(DATA_DIR).child(filename.clone());
+        let mut writer = FileWriter::<ManifestDescribing>::try_new(
+            &object_store,
+            &full_path,
+            schema,
+            &Default::default(),
+        )
+        .await?;
```
Comment (westonpace) [inline rust/lance/src/dataset/fragment/write.rs:93]: This is no different than before but I suppose we will need to update this path to use the v2 writer as well.

**Review by westonpace — APPROVED** [review]
This looks pretty cool.

File: rust/lance/src/dataset/fragment/write.rs (hunk 3 above)
Comment (wjones127) [reply rust/lance/src/dataset/fragment/write.rs:93]: Good point. I've added a TODO to https://github.com/lancedb/lance/issues/1929

**Pull request merged**